src/Sarif/Visitors/SarifCurrentToVersionOneVisitor.cs

```diff
@@ -1032,6 +1014,81 @@ private static string CreateFileDictionaryKey(FileData v2File, IList<FileData> v
             return logicalLocationsVersionOne;
         }
 
+        // In SARIF v1, run.resources.rules was a dictionary, whereas in v2 it is an array.
+        // Normally, the lookup key into the v1 rules dictionary is the rule id. But some
+        // tools allow multiple rules to have the same id. In that case we must synthesize
+        // a unique key for each rule with that id. We choose "<ruleId>-<n>", where <n> is
+        // 1 for the second occurrence, 2 for the third, and so on.
+        private static IDictionary<int, string> CreateV2RuleIndexToV1KeyMapping(IList<Rule> rules)
+        {
+            var v2RuleIndexToV1KeyMap = new Dictionary<int, string>();
+
+            if (rules != null)
+            {
+                // Keep track of how many distinct rules have each id.
+                var ruleIdToCountMap = new Dictionary<string, int>();
+
+                for (int i = 0; i < rules.Count; ++i)
+                {
+                    Rule rule = rules[i];
+                    if (rule.Id != null)
+                    {
+                        ruleIdToCountMap[rule.Id] = ruleIdToCountMap.ContainsKey(rule.Id)
+                            ? ruleIdToCountMap[rule.Id] + 1
+                            : 1;
+
+                        v2RuleIndexToV1KeyMap[i] = ruleIdToCountMap[rule.Id] == 1
+                            ? null
+                            : rule.Id + '-' + (ruleIdToCountMap[rule.Id] - 1).ToString();
+                    }
+                }
+            }
+
+            return v2RuleIndexToV1KeyMap;
+        }
+
+        private static string GetV1RuleKeyFromV2Index(
+            int ruleIndex,
+            IDictionary<int, string> v2RuleIndexToV1KeyMap)
+        {
+            string ruleKey = null;
+            if (0 <= ruleIndex && ruleIndex < v2RuleIndexToV1KeyMap?.Count)
```

Comment:
>ruleIndex < v2RuleIndexToV1KeyMap [](start = 34, length = 33)

wait, are you sure about this? v2RuleIndexToV1 is a map, and we skip rule entries with no ids. So I don't think a comparison to the map count is right.

Consider a rules array with 49 checks with no ids. the 50th does have a rule idea and we have a single SARIF result pointing to it with ruleIndex == 49. your v2RuleIndexToV1KeyMap will only have a single entry right? and so we'll fail this test. 

I don't understand all these hijinks in any case. why can't you just look for ruleIndex and try to find the key? it will be set to null in all failures (which is right). All your code assumes that the map may have values in it for things like ruleIndex <0 (which we want to avoid retrieving). why would that be?

Reply:
You're right that this is a bug. And you're right that there's no need for the condition at all. I don't know what I was thinking.

Then the question is whether to get rid of this helper altogether. It has the advantage of returning its value rather than using an `out` parameter. I think I'll keep it.

---
In reply to: [249575749](https://github.com/Microsoft/sarif-sdk/pull/1228#discussion_r249575749) [](ancestors = 249575749)